core/src/main/java/org/elasticsearch/search/slice/SliceBuilder.java

```diff
@@ -0,0 +1,249 @@
+            throw new IllegalArgumentException("id must be greater than or equal to 0");
+        }
+        if (max != -1 && id >= max) {
+            throw new IllegalArgumentException("max must be greater than id");
+        }
+        this.id = id;
+        return this;
+    }
+
+    /**
+     * The id of the slice.
+     */
+    public int getId() {
+        return id;
+    }
+
+    private SliceBuilder setMax(int max) {
+        if (max <= 0) {
+            throw new IllegalArgumentException("max must be greater than 0");
+        }
+        if (id != -1 && id >= max) {
+            throw new IllegalArgumentException("max must be greater than id");
+        }
+        this.max = max;
+        return this;
+    }
+
+    /**
+     * The maximum number of slices.
+     */
+    public int getMax() {
+        return max;
+    }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject();
+        innerToXContent(builder);
+        builder.endObject();
+        return builder;
+    }
+
+    void innerToXContent(XContentBuilder builder) throws IOException {
+        builder.field(FIELD_FIELD.getPreferredName(), field);
+        builder.field(ID_FIELD.getPreferredName(), id);
+        builder.field(MAX_FIELD.getPreferredName(), max);
+    }
+
+    public static SliceBuilder fromXContent(QueryParseContext context) throws IOException {
+        SliceBuilder builder = PARSER.parse(context.parser(), new SliceBuilder(), context);
+        return builder;
+    }
+
+    @Override
+    public boolean equals(Object other) {
+        if (!(other instanceof SliceBuilder)) {
+            return false;
+        }
+
+        SliceBuilder o = (SliceBuilder) other;
+        return ((field == null && o.field == null) || field.equals(o.field))
+            && id == o.id && o.max == max;
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(this.field, this.id, this.max);
+    }
+
+    public Query toFilter(QueryShardContext context, int shardId, int numShards) {
+        final MappedFieldType type = context.fieldMapper(field);
+        if (type == null) {
+            throw new IllegalArgumentException("field " + field + " not found");
+        }
+
+        if (UidFieldMapper.NAME.equals(field)) {
+            if (numShards == 1) {
+                return new TermsSliceQuery(field, id, max);
+            }
+
+            if (max >= numShards) {
+                // the number of slices is greater than the number of shards
+                // in such case we can reduce the number of requested shards by slice
+
+                // first we check if the slice is responsible of this shard
+                int targetShard = id % numShards;
+                if (targetShard != shardId) {
+                    // the shard is not part of this slice, we can skip it.
+                    return new MatchNoDocsQuery("this shard is not part of the slice");
```

Comment:
maybe we could improve the transport scroll action to not even request those shards? (in a follow-up PR)